From f31ffe368aa2b467739c74b03321efec60b06222 Mon Sep 17 00:00:00 2001
From: Paco Hope <pacohope@users.noreply.github.com>
Date: Wed, 17 Jun 2020 09:14:10 -0400
Subject: [PATCH 13/18] Freebsd support (#1)

Add detection and installation for FreeBSD

The tools are installed using FreeBSD's repositories directly, using the `pkg` tool.
---
 mk/install.sh            | 31 ++++++++++++++++++++++++++++---
 mk/xe-linux-distribution | 21 +++++++++++++++++++++
 2 files changed, 49 insertions(+), 3 deletions(-)

diff --git a/mk/install.sh b/mk/install.sh
index 4faae79..5ecce43 100755
--- a/mk/install.sh
+++ b/mk/install.sh
@@ -75,7 +75,7 @@ fi
 
 ARCH=$(uname -m | sed -e 's/i.86/i386/g')
 case "${ARCH}" in
-    i386|x86_64) ;;
+    i386|x86_64|amd64) ;;
     *)
 	echo "Architecture ${ARCH} is not supported"
 	exit 1
@@ -230,6 +230,21 @@ select_pkgs_xe()
     select_rpm_utilities
 }
 
+select_pkgs_freebsd()
+{
+    GUEST_PKG_TYPE=pkg
+    PKGNAME=$(pkg info xe-guest-utilities 2>&1)
+    ERROR=$?
+    if [ $ERROR != 0 ]
+    then
+        XGU="xe-guest-utilities"
+    else
+        set $PKGNAME
+        PKGNAME="$1"
+        echo "${PKGNAME} already installed"
+    fi
+}
+
 update_lvm_configuration_required()
 {
     if test "${os_distro}" != rhel
@@ -517,6 +532,14 @@ install_coreos()
     systemctl start xe-linux-distribution.service
 }
 
+install_freebsd()
+{
+    echo "Installing xe-guest-utilities from FreeBSD pkg..."
+    pkg install xe-guest-utilities
+    echo "Installation complete. Starting xenguest"
+    service xenguest start
+}
+
 case "${os_distro}" in
     rhel|centos|oracle|fedora|rocky)         select_pkgs_rhel ;;
     scientific|neokylin|asianux|turbo) select_pkgs_rhel ;;
@@ -525,6 +548,7 @@ case "${os_distro}" in
     debian|ubuntu|linx|yinhe|Gooroom)  select_pkgs_debian ;;
     xe-ddk|xe-sdk)                     select_pkgs_xe ;;
     *CoreOS)                            select_pkgs_coreos ;;
+    FreeBSD)                           select_pkgs_freebsd ;;
     *)                  failure "Unknown Linux distribution \`${os_distro}'." ;;
 esac
 
@@ -551,7 +575,7 @@ if [ -n "${ECRYPTFS_UTILS}" ] ; then
 	fi
     done
 fi
-if [ -n "${XGU}" ] ; then
+if [ -n "${XGU}" ] && [ "${os_distro}" != "FreeBSD" ] ; then
     for P in ${XGU} ; do
 	if [ ! -f "${P}" ] ; then
 	    echo "Warning: xe-guest-utilities ${P} not found."
@@ -559,7 +583,7 @@ if [ -n "${XGU}" ] ; then
 	fi
     done
 fi
-if [ -z "${XGU}" ] ; then
+if [ -z "${XGU}" ] && [ "${os_distro}" != "FreeBSD" ] ; then
     echo ""
     echo "Certain guest features will not be active until a version of "
     echo "xe-guest-utilities is installed."
@@ -677,6 +701,7 @@ case ${GUEST_PKG_TYPE} in
     rpm) install_rpms ;;
     deb) install_debs ;;
     coreos) install_coreos ;;
+    pkg) install_freebsd ;;
 esac
 
 if [ -n "${KERNEL}" -o -n "${XGU}" ] ; then
diff --git a/mk/xe-linux-distribution b/mk/xe-linux-distribution
index ee568dc..6cd22c4 100755
--- a/mk/xe-linux-distribution
+++ b/mk/xe-linux-distribution
@@ -568,7 +568,27 @@ identify_boot2docker()
     minor=$(awk -F. '{printf("%s.%s", $2, $3)}' /etc/version)
 
     write_to_output "boot2docker" "${major}" "${minor}" "boot2docker $(head -n 1 /etc/version)"
+}
+
+identify_freebsd()
+{
+
+    if [ ! -x /usr/bin/uname ] || [ $(/usr/bin/uname -s) != "FreeBSD" ]
+    then
+        return 1
+    fi
 
+    distro="FreeBSD"
+    release=$(uname -r)
+    IFS="-"
+    set - $release
+    version=$1
+    IFS="."
+    set - $version
+    major=$1
+    minor=$2
+    osname="FreeBSD $release"
+    write_to_output "${distro}" "${major}" "${minor}" "${osname}"
 }
 
 if [ $# -eq 1 ] ; then
@@ -598,6 +618,7 @@ if [ -z "${TEST}" ] ; then
     identify_debian /etc/debian_version && exit 0
     identify_boot2docker /etc/boot2docker && exit 0
     identify_sangoma /etc/centos-release && exit 0
+    identify_freebsd && exit 0
     identify_os_release /etc/os-release && exit 0
 
 
-- 
2.30.2

